### PR TITLE
Merge Docker-Setup

### PR DESCRIPTION
**Pull request by khassel**
Discussion from https://github.com/bastilimbach/docker-MagicMirror/pull/4, sources from https://github.com/khassel/docker-mm merged and updated for bastilimbach docker repo.

### CODE REVIEW

**Comment by bastilimbach** [comment]
Thank you, great work! I changed the base branch so I can test this with Docker Hub before I merge it into master

**Comment by khassel** [comment]
I will be offline for the next 10 days, so I wrote a short description how to setup the host pi. Have no time to put it into your Readme.md at the moment, so I post it here (may you can put it in the Readme.md or I will do later):

Setup for your host raspberry pi for running MagicMirror using Docker.

Requirements:
• running raspberry pi version 2 or 3 with running raspian jessie
• LAN or WLAN access
• logged in as user pi (otherwise you have to substitute "pi" with your user in the following lines)

Setup for MagicMirror
• create local directory: mkdir -p ~/magic_mirror/config 
  and add your config.js file
• create local directory: mkdir -p ~/magic_mirror/modules 
  and add your 3rd party modules

Setup Docker
• get Docker: 			curl -sSL get.docker.com | sh
• set Docker to auto-start:	sudo systemctl enable docker
• start the Docker daemon:	sudo systemctl start docker  (or reboot your pi)
• add user pi to docker group:	sudo usermod -aG docker pi

Setup docker-compose:
• execute: 			sudo apt-get purge python-pip
• execute: 			curl https://bootstrap.pypa.io/get-pip.py | sudo python
• execute: 			sudo pip install docker-compose
 
Setup for graphical desktop:
• install unclutter:		sudo apt-get install -y unclutter
• edit (here with nano):		nano /home/pi/.config/lxsession/LXDE-pi/autostart
  and insert the following lines for disabling screensaver and mouse cursor:

  @xset s noblank
  @xset s off
  @xset -dpms
  @unclutter -idle 0.1 -pi

  uncomment the existing lines, otherwise you will see the pi desktop before MagicMirror has started
• edit (here with nano):		nano ~/.bashrc
  and insert the following lines (otherwise docker has no access on the pi display):

  export DISPLAY=:0.0
  xhost +local:root	

• execute: 			sudo raspi-config
  goto "3 boot options" and choose "B2 Wait for Network at Boot"

  If not set, some modules will remaining in "load"-state because MagicMirror starts to early.

Setup docker-MagicMirror:
• execute:			git clone --depth 1 -b master https://github.com/bastilimbach/docker-MagicMirror.git ~/docker-MagicMirror

Get the docker image:
• execute:  docker pull bastilimbach/docker-magicmirror

Starting MagicMirror:
• goto: 				cd ~/docker-MagicMirror/up-raspberry
  and execute:			docker-compose up -d

  in case you want to stop it:	docker-compose down

**Comment by khassel** [comment]
updated my develop branch on the fork https://github.com/khassel/docker-MagicMirror/tree/develop fixing localtime, may you can merge it into your "new-docker-setup" branch?
Could you test the whole thing? Any problems?
Did you update the ReadMe.md? Otherwise i will do ...

**Comment by bastilimbach** [comment]
I don't have the time to look into this at the moment. The university is pretty time-consuming. I hope I can have a look at the changes next weekend. I just wanted to let you know that I didn't forget about this PR and really appreciate your work! 👍

**Comment by khassel** [comment]
Would be great to publish a running docker-setup but university has absolute priority.
Updated my fork https://github.com/khassel/docker-MagicMirror/tree/develop again with the new readme.md. Feel free to correct typos and my "german-english" ...

File: Dockerfile [hunk 1]
```diff
@@ -1,21 +1,32 @@
...
 COPY docker-entrypoint.sh /opt/magic_mirror
-RUN apt-get update \
-  && apt-get -qy install dos2unix \
-  && dos2unix docker-entrypoint.sh \
-  && chmod +x docker-entrypoint.sh
+RUN dos2unix docker-entrypoint.sh && chmod +x docker-entrypoint.sh
 
-EXPOSE $MM_PORT
-ENTRYPOINT ["/opt/magic_mirror/docker-entrypoint.sh"]
+ENTRYPOINT ["/opt/magic_mirror/docker-entry-raspberry.sh"]
```
Comment (bastilimbach) [inline Dockerfile:32]: Shouldn't this be just `docker-entrypoint.sh` instead of `docker-entry-raspberry.sh`?

File: known-issues/Issues.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+1. update Readme.md
+
+2. 3rd party modules in general:
```
Comment (bastilimbach) [inline known-issues/Issues.txt:3]: I think this point is the main problem with the implementation of the UI over Docker. If someone is trying to implement a 3rd party module, which needs another dependency, the user is required to modify the Dockerfile and build it by himself. Or we get a ton of "Add this or that dependency" as issues.

What do you think about a little section in the README which describes the way on how to use the UI with Docker. We could create a subfolder with an example Dockerfile.

File: up-raspberry/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    environment:
+#1      LD_LIBRARY_PATH: /opt/vc/lib
+      DISPLAY: unix$DISPLAY
+#2    network_mode: host
+    entrypoint: /opt/magic_mirror/docker-entrypoint.sh "npm start"   
+    restart: always
+
+#1  needed for module MMM-Systemtemperature
+#2  needed for module MMM-NetworkScanner
+#3  needed for #1 and #2
```
Comment (bastilimbach) [inline up-raspberry/docker-compose.yml:25]: With more 3rd party modules, I think this lists will get bigger and bigger. Look at the above comment.

File: Dockerfile (hunk 1 above)
Comment (khassel) [reply Dockerfile:32]: yes, the docker-entry-raspberry.sh doen't exist anymore.

The entrypoint is set in the 2 docker-compose.yml files which overrides the default entrypoint defined in the Dockerfile ...

**Comment by bastilimbach** [comment]
Another point, why I think we shouldn't implement the UI part:
If the users set up the server only mode using Docker. He can just open the raspberry browser in kiosk mode and go to localhost or whatever the docker address is.

Examples:
- [https://github.com/elalemanyo/raspberry-pi-kiosk-screen](https://github.com/elalemanyo/raspberry-pi-kiosk-screen)
- [https://medium.com/stories-from-upstatement/how-to-build-a-web-kiosk-with-a-raspberry-pi-some-cables-and-a-tv-3dc2724acaa1](https://medium.com/stories-from-upstatement/how-to-build-a-web-kiosk-with-a-raspberry-pi-some-cables-and-a-tv-3dc2724acaa1)

**Comment by khassel** [comment]
I think we have 2 usecases
a) server-only mode
b) running mm as docker-container on the pi.

At the moment there is only one Dockerfile and 2 docker-compose.yml for each usecase.

The Dockerfile contains some things only for b), this is 

libgtk2.0-0 libx11-xcb-dev libxtst6 libxss1 libgconf-2-4 libnss3-dev libasound2

for running on the pi and

arp-scan for MMM-NetworkScanner (thats because I use this module).

May we should check which 3rd-Party-modules need extra packages to decide if we should do all in one Dockerfile?

I think most users will not create own Dockerfiles, editing the docker-compose.yml for extra starting parameters is not so difficult.

**Comment by bastilimbach** [comment]
Right and I think we should stick with just the first use case because we just cannot implement all the necessary dependencies for all 3rd party modules. First, it would be pretty time consuming to keep up with all the new modules and second the docker container would increase in unnecessary size if someone doesn't need all these modules.

What does the electron app, running in Docker, do better than the standard browser from raspbian or noobs in kiosk mode? I think the setup, to auto-start the browser in kiosk mode is way easier than fiddling with the compose file every time you want to use a new module. What do you think?

**Comment by bastilimbach** [comment]
BTW the travis auto deploy (for server only) [works](https://travis-ci.org/bastilimbach/docker-MagicMirror/builds/236948692) 👍

**Comment by bastilimbach** [comment]
[Docker Hub Tags](https://hub.docker.com/r/bastilimbach/docker-magicmirror/tags/)

**Comment by khassel** [comment]
There is no real need for usecase b). It was fun to get it working. You can do a normal mm install without docker or run server-only in a browser on the pi.
But if you only implement server-only, why 3 docker tags? Why alpine and raspberry?

**Comment by bastilimbach** [comment]
The alpine version is much smaller than the latest version. Some people need to deal with limited disk space, so I wanted to offer an alternative.

**Pull request merged**